### ISSUE TITLE
[2.0] Standardize error handling & remove race conditions

### ISSUE THREAD

**Issue by SamMousa** [thephpleague/flysystem#1018]
Currently the `FileSystem` allows for race conditions (https://en.wikipedia.org/wiki/Time_of_check_to_time_of_use) in adapters to lead to unpredictable errors.

Several discussions exist around this issue: #972, #940.

Regardless of using checks before use, we should still handle errors that happen during use. I propose adding this to all functions in the adapter interfaces:
```php
/**
   This function must not trigger warnings, errors must be communicated by throwing exceptions
*/

```

The next step would be to allow adapters to throw more specific exceptions that are standardized by FlySystem. Advantage would be that we could show the user better errors without understanding or knowing about the underlying implementation; currently we have `FileNotFoundException`, but other common things like `AccessDeniedException` could be added.
Since an adapter knows most about its underlying implementation this allows it to provide better information to the developer in a documented way.
 
If we make these changes to the interface (they are PHPDoc only changes), we will be able to suppress warnings and instead convert them to exceptions. Using exceptions that can be standardized to some generic cases will allow consumers of the API to handle known exceptions in better ways and unknown exceptions via their normal error handling (like logging it and acting on it).

I know most of these discussions above have been about the `Local` adapter, where one could argue that race conditions will be rare, other adapters like the `Ftp` adapter have the same issue. Roundtrip times to an FTP server can easily be > 500ms, combining this with the fact that `Ftp` file systems are multi user file system, this makes these race conditions much more likely.

In conclusion I think we can broaden the scope of FlySystem by including guidance on how to communicate errors to the API consumers.
